src/engine/sched.c

```diff
@@ -1081,25 +1175,80 @@ process_pool_cb(d_list_t *rlink, void *arg)
 	return 0;
 }
 
-static void
+static int
 policy_fifo_enqueue(struct dss_xstream *dx, struct sched_request *req,
 		    void *prio_data)
 {
 	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_req_attr	*attr = &req->sr_attr;
+
+	D_ASSERT(attr->sra_type < SCHED_REQ_TYPE_MAX);
+	/*
+	 * If @sra_enqueue_id is not zero, this is a resent RPC, it will
+	 * be inserted to sorted heap instead of fifo list.
+	 */
+	if (attr->sra_enqueue_id)
+		return d_binheap_insert(&info->si_heap, &req->sr_node);
 
 	d_list_add_tail(&req->sr_link, &info->si_fifo_list);
+
+	return 0;
 }
 
 static void
 policy_fifo_process(struct dss_xstream *dx)
 {
 	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_request	*req, *tmp, *req1;
+	struct d_binheap_node	*node;
+	d_list_t                 tmp_list;
+	int			 rc;
 
-	process_req_list(dx, &info->si_fifo_list, false);
+	D_INIT_LIST_HEAD(&tmp_list);
+	/*
+	 * All retried RPCs are inserted into a sorted heap, they are sorted
+	 * by RPC enqueue sequence ID in the server side(firstly enqueue time). So
+	 * retried RPCs won't starve forever.
+	 */
+	d_list_for_each_entry_safe(req, tmp, &info->si_fifo_list, sr_link) {
```

Comment:
the request in si_fifo_list is new request, and the request in si_heap is retried old request?
here is to ensure if with retried request with smaller sra_enqueue_id than the new request, it will process the retried request? why must ensure this order?
If the retried request has not arrive to server, that order cannot be ensured anyway right?

Reply:
smaller ID means it arrived server earlier in the first time, it is still a kind of "FIFO" to some retried RPCs could be not handlined for a long time which might cause tail-latency very high for some application. 

For second question, yup, we don't know when retried request might arrive to server, but once arrive, smaller ID always
handle firstly to avoid above issue.